portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -39,26 +44,40 @@
   private List<Comment> arr = new ArrayList<>();
 
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
     Query commentsQuery = new Query("Comment").addSort("timestamp");
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
     PreparedQuery results = datastore.prepare(commentsQuery);
     List<Comment> arr = new ArrayList<>();
     int how_many = Integer.parseInt(request.getParameter("how_many"));
+    String langCode = (String) request.getParameter("lang");
 
+    String[] acceptedCodes = {"en", "ro", "de", "fr"};
+    if (!Arrays.asList(acceptedCodes).contains(langCode)){
+      response.setContentType("text/html");
+      response.getWriter().print("<script>alert(\"Please use an accepted language code.\")</script>");
+      RequestDispatcher dispatcher = request.getRequestDispatcher("/comments.html");
+      dispatcher.include(request, response);
+      return;
+    }
+    
     for (Entity it : results.asIterable()) {
       String message = (String) it.getProperty("message");
       String uid = (String) it.getProperty("id");
       long timestamp = (long) it.getProperty("timestamp");
       long id = (long) it.getKey().getId();
 
+      Translate translate = (Translate) TranslateOptions.getDefaultInstance().getService();
+      Translation translation = translate.translate(message, Translate.TranslateOption.targetLanguage(langCode));
```

Comment:
How is the source language handled/detected?

Reply:
I think automatically, I just followed the walkthrough and they didn't mention anything about the source language.